Enable EKS auth via AssumeRoleWithWebIdentity

fh-mthomson opened the pull request:
Closes #477 and #549

Testing: confirmed that `paws` commands "just worked" (without any credential handling) using this version of `paws.common` when deploying to RStudio Connect hosted on EKS. Truly magic, what a great package!

fh-mthomson commented on docs/credentials.md:
```diff
@@ -51,7 +51,7 @@ Sys.setenv(
 If you have a session token from temporary security credentials, you
 can set it and its expiration time in environment variables `AWS_SESSION_TOKEN`
-and `AWS_CREDENTIAL_EXPIRATION`:
+and `AWS_CREDENTIAL_EXPIRATION` (optional):
```
Addresses #549

fh-mthomson commented on docs/credentials.md:
````diff
@@ -130,14 +130,20 @@ svc <- paws::svc(
 
+Additionally, anonymous credentials may be used (e.g., before calling `assume_role_with_web_identity()` on an STS service)
+``` r
+svc <- paws::svc(config = list(credentials = list(anonymous = TRUE)))
````
This fully unblocked me once I stumbled upon the great discussion in #512, so thought it'd be helpful to pull out more visibly!

fh-mthomson commented on paws.common/R/config.R:
```diff
@@ -150,29 +150,16 @@ get_env <- function(variable) {
-    value <- "" # Not implemented on Windows.
-  }
+# Get the name of the IAM role from the instance metadata.
+get_iam_role <- function() {
```
Moved `get_iam_role()` to `config.R` from `credential_providers.R`, given parallel to `get_instance_metadata()` and similar functions in the former (e.g., pulling from env variables)

DyfanJones replied:
sounds good to me

fh-mthomson commented on paws.common/R/config.R:
```diff
@@ -150,29 +150,16 @@ get_env <- function(variable) {
 
-# Get the value of an OS environment variable.
-# NOTE: Does not work on Windows.
-get_os_env <- function(var) {
```
`get_os_env()` and `get_profile_name()` moved below (alphabetical) without touching function

DyfanJones replied:
happy for them to be moved

fh-mthomson commented on paws.common/R/credential_providers.R:
```diff
@@ -322,7 +341,28 @@ get_container_credentials <- function() {
   }
 
-  return(metadata_response)
+  credentials_response_body <-
```
`get_container_credentials()` should return... _credentials_. While this is technically changing the API, I think it's a meaningful improvement and shouldn't impact users, since these are internal functions

DyfanJones replied:
makes sense to me :)

fh-mthomson commented:
Hey @DyfanJones ! I took a pass at addressing #477 since it was the key blocker for some of our development efforts. 
Would you be able to review when you have a chance? Thank you!

DyfanJones commented:
@fh-mthomson thanks for the PR. I will be able to review this on monday (if that is ok). I am travelling home from spain on sunday, and i dont want to do a rushed review :) from initial glance it looks really good. Our initial design is base off the aws go sdk (v1), however i am sure what you have done will be ok 😄

fh-mthomson commented:
Absolutely -- safe travels, @DyfanJones! Looking forward to your feedback then :)

fh-mthomson commented:
I'd also be happy to adjust to more closely mirror the aws go sdk, I hadn't heard of that! Feel free to point me in the right direction! 

A big thank you for your efforts in this package -- it was such a good find!

DyfanJones commented:
The aws go sdk it fairly straight forward to follow compared to the other sdks i.e. boto3. Of course there are R differences however it is great point of reference. I am not saying "you must have corresponding stuff to aws go sdk" :) but it is great for the reference guide and help update the current r sdk with latest aws features (https://github.com/aws/aws-sdk-go).

DyfanJones commented:
@fh-mthomson thanks for going the extra mile and checking against aws go sdk. I have been feeling under the weather for the past couple of days, but i will get onto reviewing this shortly. Sorry for the waiting :)

fh-mthomson commented:
Totally understand, no rush! Take good care :)

DyfanJones commented on paws.common/NEWS.md:
```diff
@@ -2,6 +2,8 @@
 * Fix Content-Md5 being modified by user
 * Add logging system to help with debugging paws issues
+* Add support for `AssumeRoleWithWebIdentity` (#477)
+* Clarify that `AWS_CREDENTIAL_EXPIRATION` is optional when using `AWS_SESSION_TOKEN` (#549)
```
feel free to add yourself to this i.e. (thanks to @fh-mthomson for contribution)

fh-mthomson replied:
Added and resolved merge conflicts :)

DyfanJones commented:
@fh-mthomson sorry about the delay, I am not seeing any big issues with this currently. I will start the unit tests 😄

fh-mthomson commented:
Thanks for the review, @DyfanJones! I've pushed one minor change after rebasing but should be near final once test runs pass.

DyfanJones commented:
@fh-mthomson just a quick question around the coverage. For the new code added do you what code isn't covered in the updated unit tests?  Just wondering why the coverage has dropped a little :)

fh-mthomson commented:
> @fh-mthomson just a quick question around the coverage. For the new code added do you what code isn't covered in the updated unit tests? Just wondering why the coverage has dropped a little :)

For sure! Looking at the coverage report, seems to be all in `credential_providers.R`, which has a baseline coverage of ~5%.
<img width="685" alt="y Additional details and impacted files" src="https://user-images.githubusercontent.com/56164984/203144862-cdbec8ee-ecba-4566-964f-33631690d2e1.png">

I'll admit that I'm new to this code coverage, but this makes sense since I added several functions without tests.
```
get_creds_from_sts_resp() # -> now added a test!
get_container_credentials_eks()
get_assume_role_with_web_identity_creds()
```

For the latter two, I believe they'd require more in depth testing (e.g., mocks) which are currently outside of my wheelhouse. Let me know if you think these are crucial!

DyfanJones commented:
Ah I see, I think they are a nice to have as long as the = core functionality is covered part of other unit tests :) but thanks for updating the unit tests 😄. In general we try to keep our coverage fairly high so we can trust our testing, and any changes to the internal api can be trusted if they pass the unit tests 😄

fh-mthomson commented:
Makes total sense! Looks like the decrease is -0.6% instead of -0.8%, so a bit better :)

davidkretch commented:
@DyfanJones @fh-mthomson Looks good to me too.
